### ISSUE TITLE
Markdown breaks code folding

### ISSUE THREAD

**Issue by vijithassar** [atom/atom#16782]
### Description

Code folding does not work for Markdown code blocks [as previously reported](https://github.com/atom/atom/issues/8879) which means Atom is not a good choice for [coding in literate programming style](https://github.com/vijithassar/lit). Hoping this can be fixed now that [v1.25.0](https://github.com/atom/atom/releases/tag/v1.25.0-beta0) introduces [a new parser](https://github.com/atom/atom/pull/16299).

### Steps to Reproduce

Code cannot be folded. Markdown code blocks which 

**Expected behavior:** functions fold according to code structure

**Actual behavior:** syntax highlighting breaks and code folding is disabled

**Reproduces how often:** 100%

### Versions

Mac OS 10.13, all versions of Atom

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by maxbrunsfeld** [comment]
> Markdown code blocks which do not completely contain a function with all brackets balanced break syntax highlighting and code folding.

This is going to continue to be somewhat challenging even with Tree-sitter. I'll try to keep it in mind as we develop the Tree-sitter markdown parser and Tree-sitter's language embedding feature in general.

**Comment by Aerijo** [comment]
@maxbrunsfeld this feels more like the regex based rules (and I don’t know how to translate it to tree sitter), but isn’t it possible to mark the start of a code block, and then mark the end of it before trying to apply the other syntax. Then, after the range has been established, the other syntax can be applied to that range only. 

For example, markdown could use this by requiring an otherwise empty line with a number of backticks matching what it started with. As this will always close the markdown block, there is no fear of accidentally exiting too early.

**Comment by 50Wliu** [comment]
> For example, markdown could use this by requiring an otherwise empty line with a number of backticks matching what it started with.

It already does in the existing implementation.  The problem is that the nested rule hasn't ended yet and so it doesn't return control to language-gfm.

> and then mark the end of it before trying to apply the other syntax

first-mate regex matches are restricted to single-line matches for performance reasons, which is why this can't be done.

**Comment by Aerijo** [comment]
>The problem is that the nested rule hasn't ended yet and so it doesn't return control to language-gfm.

Yes, thats why I said it could apply the markdown first, and then go back and apply the child one. I also suggested it be forced to always look for the end pattern first, and force back control if necessary.

> first-mate regex matches are restricted to single-line matches for performance reasons

I know, but the separate grammar can be applied in parallel even (or async or whatever C++ or JavaScript lets you do), as we know it's completely bound by the range. Alternatively, the rule in your PR would accomplish it.

> Once upon a time I experimented with fixing this in atom/first-mate#90, but I don't think that PR will be finished.

I would have liked to see that PR succeed, and I'm advocating for functionality like it here early so we don't get to a situation like that again.

**Comment by Aerijo** [comment]
I think it's worth pointing out that this approach would allow arbitrarily nested syntax highlighting, as establishing the range beforehand prevents any overflow of any of the child components. For example,

````
See the following latex snippet for using the minted package:

```latex
This is scoped as \LaTeX now

\begin{minted}{python}
  # and here we are using python
  def my_func(n):
    """ unfinished docstring
\end{minted}

But we are back to \LaTeX here, because the LaTeX parser only gave the limited range to the python one.
```

And finally, back to markdown
````
Markdown would mark the contents of the backticks as LaTeX. The LaTeX parser would then only have access to these contents, and would mark the contents of the environment as python. Even though the python string is unfinished, it is restricted to it's given range.

Having said that, I don't think much effort needs to be put into making it work. Ideally, it would just happen as a side effect of the actual child parsing function (by letting it recursively apply itself). Practically, it's unlikely many (or any) people would need it.

**Comment by vijithassar** [comment]
Still an issue. All versions, all operating systems.
